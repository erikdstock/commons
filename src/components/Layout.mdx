---
Layout
---
import { Flex } from './Layout'
import { H2 } from './Typography'

# Layout

<Flex py={3}  bg="red" justifyContent="center" color="white">
  <H2 fontFamily="mono" color="white">TODO</H2>
</Flex>