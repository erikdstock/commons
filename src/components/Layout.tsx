import { Box, Flex } from "rebass"
import styled from "styled-components"
import {
  borderColor,
  BorderColorProps,
  BorderProps,
  borders,
  color,
  space,
} from "styled-system"

const BorderBox = styled(Box)<BorderProps & BorderColorProps>`
  ${color};
  ${space};
  ${borders};
  ${borderColor};
  border-width: 2px;
  border-style: solid;
`

export { BorderBox, Box, Flex }
