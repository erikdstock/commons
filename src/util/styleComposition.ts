import styled from "@emotion/styled"
import ssCss, { CSSObject, SystemStyleObject } from "@styled-system/css"

import get from "lodash/get"
import { RosesTheme } from "../theme"

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

interface ComposeOptions {
  defaultVariant?: string
}

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | undefined

/** Reach into the theme, get a style object, apply it with styledCss */
const themed: (path: string) => InterpolationFn = path => props => {
  console.warn(path, props.theme)
  const componentStyle = get(props.theme, path)
  console.warn("Applying Style: ", componentStyle)
  return componentStyle && styledCss(componentStyle)
}

const themedComponent: (
  name: string
) => (p: StyledInterpolationProps) => CSSObject | undefined = (
  name: string
) => {
  console.warn("wrapping component with " + name + " styles")
  return themed(`${componentStylesRoot}.${name}`)
}

const rxHandler: InterpolationFn = ({ rx }) => rx && styledCss(rx)

type VariantHandler = (key: string, op: ComposeOptions) => InterpolationFn

const variantHandler: VariantHandler = (componentKey, ops) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || ops.defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    console.warn("variant: ", variantCss)
    return variantCss && styledCss(variantCss)
  }
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `css` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 * ```tsx
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *  const Widget = withStyleProps('Box', RawBox, {})
 * ```
 */
// export const withStyleProps = <P extends RosesStyleProps>(
export const withStyleProps = (
  componentKey: string,
  Component: React.ComponentType<RosesStyleProps>,
  options: ComposeOptions = {}
) => {
  return styled(Component)<RosesStyleProps>(
    themedComponent(componentKey),
    variantHandler(componentKey, options),
    rxHandler
  )
}
// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
