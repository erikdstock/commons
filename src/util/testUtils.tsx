import React from "react"
import renderer from "react-test-renderer"
import { RosesTheme } from "../Theme"

export const render = (tree: React.ReactNode) =>
  renderer.create(<RosesTheme>{tree}</RosesTheme>)
