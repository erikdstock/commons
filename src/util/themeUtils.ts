// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
