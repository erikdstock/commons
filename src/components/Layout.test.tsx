import React from "react"
import { render } from "../util/testUtils"
import { Box, Flex } from "./Layout"

describe("Box", () => {
  const tree = render(<Box>Click me</Box>).toJSON()
  it("renders correctly", () => {
    expect(tree).toMatchSnapshot()
  })
})
describe("Flex", () => {
  const tree = render(<Flex>Click me</Flex>).toJSON()
  it("renders correctly", () => {
    expect(tree).toMatchSnapshot()
  })
})
