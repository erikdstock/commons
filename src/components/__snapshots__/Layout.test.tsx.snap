// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
}

<div
  className="emotion-0"
>
  Click me
</div>
`;

exports[`Flex renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="emotion-0"
>
  Click me
</div>
`;
